### PR TITLE
Added: Support for TF State V3

### LINKED ISSUE #74: Port State version 3
Since https://github.com/cycloidio/inframap/issues/72 we found out that we can only support TFState version 4.

Version 3 also has the `depends_on` so we could manually parse the State (JSON) and build the dependencies and interpolate them instead of using the internal TF logic to do it.

Version 3 is from TF 0.11.

### PR DESCRIPTION
**Pull request by xescugc**
This PR closes #74 but in the process also found other bugs thanks to the state I used.

The main fixes are:

* mutate function now checks more things before choosing connections to merge/replace:
  - Popularity of the first node, by counting the edges it has (more edges more popular)
  - Ignoring connections from Nodes that already have direct connections, execpt if it's the only one missing
  - The shortest connection to a Node now also returns more than 1 element so if there are 3 shortest connections, all of them would be considered instead of the last one.
* Elements from `provider.IM` are not counted for the connections and cannot be connected together.

Ran the tests 100 times and they work perfectly for all the other cases we have.

### CODE REVIEW

File: generate/state.go [hunk 1]
```diff
@@ -183,23 +193,92 @@ func FromState(tfstate json.RawMessage, opt Options) (*graph.Graph, map[string]i
...
+// validateTFStateVersion validates that the version is the
+// one we support which is only 3 and 4
 func validateTFStateVersion(b []byte) error {
+	v, err := tfStateVersion(b)
+	if err != nil {
+		return fmt.Errorf("could not read version: %w", err)
+	}
+
+	if v != 4 && v != 3 {
+		return fmt.Errorf("could not read version %d: %w", v, errcode.ErrInvalidTFStateVersion)
```
Comment (xlr-8) [inline generate/state.go:265]: Shouldn't we precise like: "only version X & Y are supported" to be more explicit?

File: generate/state_test.go [hunk 2]
```diff
@@ -225,6 +225,113 @@ func TestFromState_AWS(t *testing.T) {
 		require.NotNil(t, cfg)
 		assert.Len(t, g.Nodes, 2)
 	})
+	t.Run("Version3", func(t *testing.T) {
+		src, err := ioutil.ReadFile("./testdata/version_3_state.json")
```
Comment (xlr-8) [inline generate/state_test.go:229]: Is it worth checking a file without modules too?

**Review by xlr-8 — COMMENTED** [review]
Couple questions

File: generate/state.go (hunk 1 above)
Comment (xescugc) [reply generate/state.go:265]: ok I've added it but to the `ErrInvalidTFStateVersion` as it has already the information for that.

File: generate/state_test.go (hunk 2 above)
Comment (xescugc) [reply generate/state_test.go:229]: I would say it's not needed, if there are 0 modules it'll have a `""` module (found it while fixing a bug on IM) so there are "always" modules.

But if you have an example I can add it to the test, more tests better haha.

**Review by xlr-8 — COMMENTED** [review]
LGTM

**Pull request merged**